src/flint/types/fmpq_mpoly.pyx

```diff
@@ -0,0 +1,600 @@
+from flint.flintlib.fmpq_mpoly_factor cimport fmpq_mpoly_factor_t, \
+    fmpq_mpoly_factor_init, fmpq_mpoly_factor, fmpq_mpoly_factor_clear, \
+    fmpq_mpoly_factor_squarefree
+
+cdef extern from *:
+    """
+    /* An ugly hack to get around the ugly hack of renaming fmpq to avoid a c/python name collision */
+    typedef fmpq fmpq_struct;
+    """
+
+
+cimport cython
+cimport libc.stdlib
+
+cdef dict _fmpq_mpoly_ctx_cache = {}
+
+
+@cython.auto_pickle(False)
+cdef class fmpq_mpoly_ctx(flint_mpoly_context):
+    """
+    A class for storing the polynomial context
+
+    :param nvars: The number of variables in the ring
+    :param ordering:  The term order for the ring
+    :param names:  A tuple containing the names of the variables of the ring.
+
+    Do not construct one of these directly, use `fmpz_mpoly_ctx.get_context`.
+    """
+
+    _ctx_cache = _fmpq_mpoly_ctx_cache
+
+    def __init__(self, slong nvars, ordering, names):
+        if ordering == "lex":
+            fmpq_mpoly_ctx_init(self.val, nvars, ordering_t.ORD_LEX)
+        elif ordering == "deglex":
+            fmpq_mpoly_ctx_init(self.val, nvars, ordering_t.ORD_DEGLEX)
+        elif ordering == "degrevlex":
+            fmpq_mpoly_ctx_init(self.val, nvars, ordering_t.ORD_DEGREVLEX)
+        else:
+            raise ValueError("Unimplemented term order %s" % ordering)
+
+        super().__init__(nvars, names)
+
+    cpdef slong nvars(self):
+        """
+        Return the number of variables in the context
+
+            >>> ctx = fmpq_mpoly_ctx.get_context(4, "lex", 'x')
+            >>> ctx.nvars()
+            4
+        """
+        return self.val.zctx.minfo.nvars
+
+    cpdef ordering(self):
+        """
+        Return the term order of the context object.
+
+            >>> ctx = fmpq_mpoly_ctx.get_context(4, "deglex", 'w')
+            >>> ctx.ordering()
+            'deglex'
+        """
+        if self.val.zctx.minfo.ord == ordering_t.ORD_LEX:
+            return "lex"
+        if self.val.zctx.minfo.ord == ordering_t.ORD_DEGLEX:
+            return "deglex"
+        if self.val.zctx.minfo.ord == ordering_t.ORD_DEGREVLEX:
+            return "degrevlex"
+
+    def gen(self, slong i):
+        """
+        Return the `i`th generator of the polynomial ring
+
+            >>> ctx = fmpq_mpoly_ctx.get_context(3, 'degrevlex', 'z')
+            >>> ctx.gen(1)
+            z1
+        """
+        cdef fmpq_mpoly res
+        assert i >= 0 and i < self.val.zctx.minfo.nvars
+        res = fmpq_mpoly.__new__(fmpq_mpoly)
+        res.ctx = self
+        fmpq_mpoly_init(res.val, res.ctx.val)
+        res._init = True
+        fmpq_mpoly_gen(res.val, i, res.ctx.val)
+        return res
+
+    def constant(self, z):
+        """
+        Create a constant polynomial in this context
+        """
+        cdef fmpq_mpoly res
+        z = any_as_fmpq(z)
+        if z is NotImplemented:
+            raise ValueError("A constant fmpq_mpoly is a fmpq")
```

Comment:
This should probably be TypeError. Is there a precedent for a `constant` method like this elsewhere?

Reply:
I agree, I think this is also one I haven't changed in fmpz_mpoly as well